agcaballero/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -33,68 +54,136 @@
-    // add available times to times array
-    for(int i = 0; i < minutes.length; i++) {
-      // if this is part of an available time range
-      if(available) {
-        // then, if current minute is false or you've reached end of day, add a new time range
-        if(! minutes[i]  || i == MINUTES_IN_DAY) {
-          int end = i;
-          int duration = end - start;
+    switch (minutes[start]) {
+      case ALL_AVAILABLE:
+       availableWithOptional = true;
+      case MANDATORY_AVAILABLE: 
+       available = true;
+    }
 
-          if(duration >= request.getDuration())
-            times.add(TimeRange.fromStartEnd(start, end - 1, true)); // add time range (inclusive of start & end) 
+    // go through minutes array once and find all the available times with optional attendees 
+    // & without optional attendees. for each availability status (all available, mandatory available,
+    // and none available), you have to either add an available time or start an availability period
+    // for other the times (with just mandatory) or timesWithOptional (all attendees).
+    for (int i = 0; i < minutes.length; i++) {
+      switch (minutes[i]) {
+        case ALL_AVAILABLE:
+          // start availability period for all attendees (including optional)
+          if (!availableWithOptional) {
+            startWithOptional = i;
+            availableWithOptional = true;
+          }
 
-          available = false;
-        }
-            
-      }
-      // if current time has been taken until now
-      else {
-        // if now available, set start to now & available to true
-        if(minutes[i]) {
-          start = i;
-          available = true;
-        }
+          // start (or continue) availability period for just mandatory atttendees
+          if (!available) {
+            start = i;
+            available = true;
+          }
+
+          break;
+
+        case MANDATORY_AVAILABLE:
+          // end availability period for all attendees & add a possible meeting time range
+          if (availableWithOptional) {
+            addMeeting(request, startWithOptional, i, false, timesWithOptional);
+            availableWithOptional = false;
+          }
+          
+          // start (or continue) availability period for just mandatory atttendees
+          if (!available) {
+            start = i;
+            available = true;
+          }
+
+          break;
+
+        case UNAVAILABLE: 
+          // end availability period for all attendees & add a possible meeting time range
+          if (availableWithOptional) {
+            addMeeting(request, startWithOptional, i, false, timesWithOptional);
+            availableWithOptional = false;
+          }
+          
+          // end availability period for mandatory attendees & add a possible meeting time range
+          if (available) {
+            addMeeting(request, start, i, false, times);
+            available = false;
+          }
+
+          break;
       }
     }
```

Comment:
Two concerns:
1. there's a fair bit of duplicate code here, I'm wondering if there are ways to reduce the duplication (see the DRY principle: https://en.wikipedia.org/wiki/Don%27t_repeat_yourself);
2. this ends up creating both arrays, even if only one of them is needed; perhaps unnecessary processing could be avoided.

One way may be to encapsulate this block of code within a method that accepts a set of Availability values, like:
```
List<TimeRange> timeRangesFromMinuteAvailabilities(Availability[] minuteAvailabilities, Set<Availability> matchingAvailabilities);
```

Then instead of `switch (minutes[i])`, you'd do `matchingAvailabilities.contains(minuteAvailabilites[i])`.

And below, you could then avoid generating the second array if not needed.

```
List<TimeRange> availableTimeRangesIncludingOptionalAttendees =
    timeRangesFromMinuteAvailabilities(minutes, EnumSet.of ( ALL_AVAILABLE ));
if (availableTimeRangesIncludingOptionalAttendees.size() > 0) {
  return availableTimeRangesIncludingOptionalAttendees;
}
return timeRangesFromMinuteAvailabilities(minutes, EnumSet.of ( ALL_AVAILABLE, MANDATORY_AVAILABLE ));
```

Reply:
I made these changes. I do think it's good to point out that this slightly changes the efficiency of the method. It now makes potentially two sweeps of the minutes array as opposed to one. However, since this doesn't change the time complexity (and this updated change will potentially save time if there is a time when all attendees can make it by only building the one array), I think it's fair to say the trade-off in readability and reduction of redundancy is worth the change.

Reply:
Done